cmd/agent/subcommands/snmp/command.go

```diff
@@ -312,11 +385,88 @@ func newSNMP(connParams *connectionParams) (*gosnmp.GoSNMP, error) {
 		MsgFlags:                msgFlags,
 		SecurityParameters:      securityParams,
 		UseUnconnectedUDPSocket: connParams.UseUnconnectedUDPSocket,
+		Logger:                  gosnmp.NewLogger(snmplog.New(logger)),
 	}, nil
 }
 
+func scanDevice(connParams *connectionParams, args argsType, conf config.Component, logger log.Component, demux demultiplexer.Component) error {
+	// Parse args
+	if len(args) == 0 {
+		return confErrf("missing argument: IP address")
+	}
+	deviceAddr := args[0]
+	if len(args) > 1 {
+		return confErrf("unexpected extra arguments; only one argument expected.")
+	}
+	// Parse port from IP address
+	connParams.IPAddress, connParams.Port, _ = maybeSplitIP(deviceAddr)
+	agentErr := setDefaultsFromAgent(connParams, conf)
+	if agentErr != nil {
+		// Warn that we couldn't contact the agent, but keep going in case the
+		// user provided enough arguments to do this anyway.
+		fmt.Fprintf(os.Stderr, "Warning: %v\n", agentErr)
+	}
+	// Establish connection
+	snmp, err := newSNMP(connParams, logger)
+	if err != nil {
+		// newSNMP only returns config errors, so any problem is a usage error
+		return configErr{err}
+	}
+	if err := snmp.Connect(); err != nil {
+		return fmt.Errorf("unable to connect to SNMP agent on %s:%d: %w", snmp.LocalAddr, snmp.Port, err)
+	}
+	defer snmp.Conn.Close()
+	pdus, err := gatherPDUs(snmp)
+	if err != nil {
+		return err
+	}
+
+	namespace := conf.GetString("network_devices.namespace")
+	deviceID := namespace + ":" + deviceAddr
+	var deviceOids []*metadata.DeviceOID
+	for _, pdu := range pdus {
+		record, err := metadata.DeviceOIDFromPDU(deviceID, pdu)
+		if err != nil {
+			logger.Warnf("PDU parsing error: %v", err)
+			continue
+		}
+		deviceOids = append(deviceOids, record)
+	}
+	sender, err := demux.GetDefaultSender()
+	if err != nil {
+		return fmt.Errorf("unable to get sender: %w", err)
+	}
+	metadataPayloads := metadata.BatchDeviceScan(namespace, time.Now(), metadata.PayloadMetadataBatchSize, deviceOids)
+	for _, payload := range metadataPayloads {
+		payloadBytes, err := json.Marshal(payload)
+		if err != nil {
+			logger.Errorf("Error marshalling device metadata: %v", err)
+			continue
+		}
+		logger.Debugf("Device OID metadata payload is %d bytes", len(payloadBytes))
+		logger.Tracef("Device OID metadata payload: %s", string(payloadBytes))
+		sender.EventPlatformEvent(payloadBytes, eventplatform.EventTypeNetworkDevicesMetadata)
```

Comment:
I'm not sure how much of a concern this is but EventPlatformEvent happens asynchronously and it buffers data up to a certain point. With such a for loop, there is a chance that we fill this buffer up for large device scan.

One possible improvement is to use `SendEventPlatformEventBlocking` from the aggregator instead

Reply:
Is there documentation for these methods somewhere? I used `EventPlatformEvent` because that's what Alex's prototype used, but the only actual documentation for this method I could find is the single sentence ["Event submits an event"](https://github.com/DataDog/datadog-agent/blob/b40a56ec8775a013ec7158fe9315bd93d6a9eeff/pkg/aggregator/sender.go#L379-L380)

I assumed, since the implementation of `EventPlatformEvent` just wrote the packet to a channel, that if the channel was full it would block, but looking at the code now I guess it gets pulled from the channel immediately and fed to `forwarder.SendEventPlatformEvent`? I see that if the buffer is full `SendEventPlatformEvent` returns an error instead of sending; just for future reference, is there any way I could've gotten that error when calling `sender.EventPlatformEvent`?

It looks like to use `forwarder.SendEventPlatformEventBlocking` I need a `message.Message`, and `EventPlatformEvent` just uses `message.NewMessage(payloadBytes, nil, "", 0)` - is there any reason I should set origin/status/ingestionTimestamp to anything other than `nil/""/0`?

Reply:
> is there any way I could've gotten that error when calling sender.EventPlatformEvent?

Looking at the code I don't think this is possible, the error is caught and logged but the calling method has no way of knowing if the message was correctly sent.

> is there any reason I should set origin/status/ingestionTimestamp to anything other than nil/""/0?
No, these additional fields aren't used in our track. I think logs uses them but I'm not sure about others.